codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoder.java

```diff
@@ -307,38 +338,87 @@ private Http2Stream requireStream(int streamId) {
-     * the passed promise is not completed until last frame write.
-     * </p>
-     */
-    private final class FlowControlledData extends FlowControlledBase {
-        private ByteBuf data;
-        private int size;
-
-        private FlowControlledData(ChannelHandlerContext ctx, Http2Stream stream, ByteBuf buf, int padding,
-                                    boolean endOfStream, ChannelPromise promise) {
-            super(ctx, stream, padding, endOfStream, promise);
-            this.data = buf;
-            size = data.readableBytes() + padding;
+    private final class FlowControlledFileRegion extends FlowControlledData<FileRegion> {
+        
+        private FlowControlledFileRegion(ChannelHandlerContext ctx, Http2Stream stream,
+                FileRegion data, int padding, boolean endOfStream, ChannelPromise promise) {
+            super(ctx, stream, data, (int) data.count(), padding, endOfStream, promise);
         }
 
         @Override
-        public int size() {
-            return size;
+        public void write(int allowedBytes) {
+            int bytesWritten = 0;
+            if (data == null || (allowedBytes == 0 && size != 0)) {
+                // No point writing an empty DATA frame, wait for a bigger allowance.
+                return;
+            }
+            try {
+                int maxFrameSize = frameWriter().configuration().frameSizePolicy().maxFrameSize();
+                do {
+                    int allowedFrameSize = Math.min(maxFrameSize, allowedBytes - bytesWritten);
+                    if (data == null) {
+                        // write padding only
+                        int writeablePadding = Math.min(allowedFrameSize, padding);
+                        padding -= writeablePadding;
+                        bytesWritten += writeablePadding;
+                        ChannelPromise writePromise;
+                        if (size == bytesWritten && !promise.isVoid()) {
+                            // Can use the original promise if it's the last write
+                            writePromise = promise;
+                        } else {
+                            // Create a new promise and listen to it for failure
+                            writePromise = ctx.newPromise();
+                            writePromise.addListener(this);
+                        }
+                        frameWriter()
+                                .writeData(ctx, stream.id(), Unpooled.EMPTY_BUFFER,
+                                        writeablePadding, size == bytesWritten && endOfStream,
+                                        writePromise);
+                        return;
+                    }
+                    FileRegion toWrite;
+                    int writeableData = (int) data.count();
+                    if (writeableData > allowedFrameSize) {
+                        FileRegion[] regions = data.split(allowedFrameSize);
```

Comment:
But finally we need to write the FileRegion out in DefaultHttp2FrameWriter... The underlying NioChannel does not take the length, it will transfer the whole FileRegion...
So maybe we need to introduce a SliceableFileRegion?